repository/src/test/java/HostedRepositoryTest.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright 2017 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import java.io.IOException;
+import java.io.InputStream;
+
+import java.net.HttpURLConnection;
+import java.net.URI;
+import java.net.URISyntaxException;
+
+import java.nio.charset.StandardCharsets;
+
+import java.util.Map;
+import java.util.Properties;
+
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.server.ServerConnector;
+import org.eclipse.jetty.server.handler.HandlerList;
+import org.eclipse.jetty.servlet.DefaultServlet;
+import org.eclipse.jetty.servlet.ServletContextHandler;
+import org.eclipse.jetty.servlet.ServletHolder;
+import org.eclipse.jetty.util.IO;
+
+import org.jbpm.process.workitem.WorkDefinitionImpl;
+import org.jbpm.process.workitem.WorkItemRepository;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import static org.hamcrest.CoreMatchers.anyOf;
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.*;
+
+public class HostedRepositoryTest {
+
+    private Server server;
+    private ServerConnector connector;
+    private Properties configProperties;
+
+    private static int PORT = 0;
+    private static String DEFAULT_HOST = "localhost";
+
+    @Before
+    public void setUp() {
+        try {
+            // load properties
+            InputStream inProps = this.getClass().getResourceAsStream("testconfig.properties");
+            configProperties = new Properties();
+            configProperties.load(inProps);
+
+            // uncomment this and change builddir and version
+            // to debug locally in IDE
+            //setConfigPropertiesForDebugging();
+
+            server = new Server();
+            connector = new ServerConnector(server);
+            connector.setPort(PORT);
```

Comment:
Unless you need to have a fresh instance every time, I'd do socket operations in `@BeforeClass`/`@AfterClass`. That way, you will minimize the number of times that the port will be opened and closed, and therefore you limit the number of times the test can potentially fail.

Reply:
@triceo do you happen to have example of this that i could reuse? Thanks

Reply:
@triceo ahh never mind i misunderstood the first time. will do

Reply:
Off the top of my head, it would look something like this:

    private static Server server;
    private static ServerConnector connector;
    
    @BeforeClass
    public static void startServer() throws Exception {
        server = new Server();
        connector = new ServerConnector(server);
        connector.setPort(PORT);
        // TODO add all that handler code in here
        server.start();
    }

    @AfterClass
    public static void stopServer() throws Exception {
        server.stop();
    }

Essentially, it's just about sprinkling a few `static`s here in there. Fair warning: I didn't actually try to run this, but I don't see why it shouldn't.